src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +25,649 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+            HashSet<INamedTypeSymbol> result = null;
+            foreach (var symbolInfo in declarationInfo.DeclaredSymbolInfos)
+            {
+                result = await ProcessSymbolInfo(
+                    document, symbolInfo,
+                    typesToSearchFor,
+                    inheritanceInfo, cachedModels,
+                    typeImmediatelyMatches, result, cancellationToken).ConfigureAwait(false);
+            }
 
-            // Must contain non-nested named types.
-            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Key.ContainingType == null));
-            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Value.ContainingType == null));
+            return result;
+        }
 
-            var referencedAssemblies = new MultiDictionary<string, IAssemblySymbol>();
-            foreach (var assembly in compilation.GetReferencedAssemblySymbols())
+        private static async Task<HashSet<INamedTypeSymbol>> ProcessSymbolInfo(
+            Document document,
+            DeclaredSymbolInfo info,
+            HashSet<INamedTypeSymbol> typesToSearchFor,
+            InheritanceInfo inheritanceInfo,
+            ConcurrentSet<SemanticModel> cachedModels,
+            Func<HashSet<INamedTypeSymbol>, INamedTypeSymbol, bool> typeImmediatelyMatches,
+            HashSet<INamedTypeSymbol> result,
+            CancellationToken cancellationToken)
+        {
+            // If we're searching for enums/structs/delegates, then we can just look at the kind of
+            // the info to see if we have a match.
+            if ((inheritanceInfo.DerivesFromSystemEnum && info.Kind == DeclaredSymbolInfoKind.Enum) ||
+                (inheritanceInfo.DerivesFromSystemValueType && info.Kind == DeclaredSymbolInfoKind.Struct) ||
+                (inheritanceInfo.DerivesFromSystemMulticastDelegate && info.Kind == DeclaredSymbolInfoKind.Delegate))
             {
-                referencedAssemblies.Add(assembly.Name, assembly);
+                var symbol = await ResolveAsync(document, info, cachedModels, cancellationToken).ConfigureAwait(false) as INamedTypeSymbol;
+                if (symbol != null)
+                {
+                    result = result ?? new HashSet<INamedTypeSymbol>();
+                    result.Add(symbol);
+                }
             }
-
-            int verifiedCount = 0;
-            foreach (var kvp in equivalentTypesWithDifferingAssemblies)
+            else if (inheritanceInfo.DerivesFromSystemObject && info.Kind == DeclaredSymbolInfoKind.Class)
             {
-                if (!verifiedKeys.Contains(kvp.Key))
+                // Searching for types derived from 'Object' needs to be handled specially.
+                // There may be no indication in source what the type actually derives from.
+                // Also, we can't just look for an empty inheritance list.  We may have 
```

Comment:
Can we still short circuit if no inheritance lists listed anything? My gut is saying that the "random standalone helper type" is common enough.